src/check_nesting.cpp

```diff
@@ -6,52 +6,413 @@
+      this->is_invalid_prop_child(node)
+    ) { return false; }
+
+    if (
+      dynamic_cast<Return*>(node) &&
+      this->is_invalid_return_parent(this->parent)
+    ) { return false; }
+
+    return true;
   }
 
-  Statement* CheckNesting::operator()(Declaration* d)
+  bool CheckNesting::is_invalid_content_parent(Statement* parent)
   {
-    if (!is_valid_prop_parent(parent())) {
-      throw Exception::InvalidSass(d->pstate(), "Properties are only allowed "
-        "within rules, directives, mixin includes, or other properties.");
+    if (!this->current_mixin_definition) {
+      throw Exception::InvalidSass(
+        parent->pstate(),
+        "@content may only be used within a mixin."
+      );
     }
-    return static_cast<Statement*>(d);
+
+    return false;
   }
 
-  Statement* CheckNesting::fallback_impl(AST_Node* n)
+  bool CheckNesting::is_invalid_charset_parent(Statement* parent)
   {
-    return static_cast<Statement*>(n);
+    if (!(
+        is_root_node(parent)
+    )) {
+      throw Exception::InvalidSass(
+        parent->pstate(),
+        "@charset may only be used at the root of a document."
+      );
+    }
+
+    return false;
   }
 
-  bool CheckNesting::is_valid_prop_parent(AST_Node* p) 
+  bool CheckNesting::is_invalid_extend_parent(Statement* parent)
   {
-    if (Definition* def = dynamic_cast<Definition*>(p)) {
-      return def->type() == Definition::MIXIN;
+    if (!(
+        dynamic_cast<Ruleset*>(parent) ||
+        dynamic_cast<Mixin_Call*>(parent) ||
+        is_mixin(parent)
+    )) {
+      throw Exception::InvalidSass(
+        parent->pstate(),
+        "Extend directives may only be used within rules."
+      );
     }
 
-    return dynamic_cast<Ruleset*>(p) ||
-           dynamic_cast<Keyframe_Rule*>(p) ||
-           dynamic_cast<Propset*>(p) ||
-           dynamic_cast<Directive*>(p) ||
-           dynamic_cast<Mixin_Call*>(p);
+    return false;
+  }
+
+  // bool CheckNesting::is_invalid_import_parent(Statement* parent)
+  // {
+  //   for (auto pp : this->parents) {
+  //     if (
+  //         dynamic_cast<Each*>(pp) ||
+  //         dynamic_cast<For*>(pp) ||
+  //         dynamic_cast<If*>(pp) ||
+  //         dynamic_cast<While*>(pp) ||
+  //         dynamic_cast<Trace*>(pp) ||
+  //         dynamic_cast<Mixin_Call*>(pp) ||
+  //         is_mixin(pp)
+  //     ) {
+  //       throw Exception::InvalidSass(
+  //         parent->pstate(),
+  //         "Import directives may not be defined within control directives or other mixins."
+  //       );
+  //     }
+  //   }
+
+  //   if (this->is_root_node(parent)) {
+  //     return false;
+  //   }
+
+  //   if (false/*n.css_import?*/) {
+  //     throw Exception::InvalidSass(
+  //       parent->pstate(),
+  //       "CSS import directives may only be used at the root of a document."
+  //     );
+  //   }
+
+  //   return true;
+  // }
+
+  bool CheckNesting::is_invalid_mixin_definition_parent(Statement* parent)
+  {
```

Comment:
Made them `void` and removed the `is_` prefix. This makes the method names the same as Ruby Sass.